src/mysql_shell.py

```diff
@@ -0,0 +1,131 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""MySQL Shell in Python execution mode
+
+https://dev.mysql.com/doc/mysql-shell/8.0/en/
+"""
+
+import dataclasses
+import json
+import logging
+import secrets
+import string
+
+import ops
+
+_PASSWORD_LENGTH = 24
+logger = logging.getLogger(__name__)
+
+
+@dataclasses.dataclass(kw_only=True)
+class Shell:
+    """MySQL Shell connected to MySQL cluster"""
+
+    _container: ops.Container
+    username: str
+    _password: str
+    _host: str
+    _port: str
+
+    _TEMPORARY_SCRIPT_FILE = "/tmp/script.py"
+
+    def _run_commands(self, commands: list[str]) -> None:
+        """Connect to MySQL cluster and run commands."""
```

Comment:
I would say, for everything except `remove_router_metadata` we should connect to mysql directly (without mysql-shell).
In general, it will be more stable.
separate PR is ok